apps/examples/rtc_alarm/rtc_alarm_main.c

```diff
@@ -0,0 +1,353 @@
+				}
+				break;
+			case 'r': 
+				/* -r: Select alarm id, if alarm is set,then it is
+				 *  done by relative time (RTC_SET_RELATIVE ioctl call). 
+				 */
+				alarmid = strtol(optarg, NULL, 0);
+				if (alarmid < 0 || alarmid >= CONFIG_RTC_NALARMS) { 
+					badarg = true;
+				}
+				relset = true;
+				break;
+			case 'c':
+				cancelmode = true;
+				break;
+			default:
+				fprintf(stderr, "<unknown parameter '-%c'>\n\n", opt);
+				/* fall through */
+
+			case '?':
+			case ':':
+				badarg = true;
+		}
+	}
+
+	setmode = !cancelmode;
+	if (setmode) {
+		if (optind >= argc) {
+			badarg = true;
+		}
+	}
+	else if (optind < argc) {
+		badarg = true;
+	}
+
+	if (badarg) {
+		show_usage(argv[0]);
+		return EXIT_FAILURE;
+	}
+
+	if (setmode) {
+		/* Get the number of seconds until the alarm expiration */
+
+		seconds = strtoul(argv[optind], NULL, 10);
+		if (seconds < 1) {
+			fprintf(stderr, "ERROR: Invalid number of seconds: %lu\n", seconds);
+			show_usage(argv[0]);
+			return EXIT_FAILURE;
+		}
+	}
+
+	/* Open the RTC driver */
+
+	printf("Opening %s\n", CONFIG_EXAMPLES_RTC_ALARM_DEVPATH);
+
+	fd = open(CONFIG_EXAMPLES_RTC_ALARM_DEVPATH, O_WRONLY);
+	if (fd < 0) {
+		int errcode = errno;
+		fprintf(stderr, "ERROR: Failed to open %s: %d\n", CONFIG_EXAMPLES_RTC_ALARM_DEVPATH, errcode);
+		return EXIT_FAILURE;
+	}
+
+	if (cancelmode) {
+		ret = ioctl(fd, RTC_CANCEL_ALARM, (unsigned long)alarmid);
+		if (ret < 0) {
+			int errcode = errno;
+			fprintf(stderr, "ERROR: RTC_CANCEL_ALARM ioctl failed: %d\n", errcode);
+			close(fd);
+			return EXIT_FAILURE;
+		}
+
+		printf("Alarm %d has been canceled\n", alarmid);
+	}
+
+	if (setmode && relset) {
+		struct rtc_setrelative_s setrel;
+
+		/* Set the alarm */
+
+		setrel.id = alarmid;
+		setrel.pid = g_alarm_daemon_pid;
+		setrel.reltime = (time_t)seconds;
+		setrel.signo = CONFIG_EXAMPLES_RTC_ALARM_SIGNO;
+		setrel.sigvalue.sival_int = alarmid;
+
+		ret = ioctl(fd, RTC_SET_RELATIVE, (unsigned long)((uintptr_t)&setrel));
+		if (ret < 0) {
+			int errcode = errno;
+			fprintf(stderr, "ERROR: RTC_SET_RELATIVE ioctl failed: %d\n", errcode);
+			close(fd);
+			return EXIT_FAILURE;
+		}
+		printf("Alarm %d set in %lu seconds\n", alarmid, seconds);
+	}
+	else if(setmode) {
+		struct rtc_setalarm_s setalarm;
+
+		/* Set the alarm */
+
+		setalarm.id = alarmid;
+		setalarm.pid = g_alarm_daemon_pid;
+
+		ret = ioctl(fd, RTC_RD_TIME, (unsigned long)&setalarm.time);
+		if (ret < 0) {
+			int errcode = errno;
+			fprintf(stderr, "ERROR: RTC_RD_ALARM ioctl failed: %d\n", errcode);
+			close(fd);
+			return EXIT_FAILURE;
+		}
+		if(setalarm.time.tm_sec + seconds >= 60) {
+			setalarm.time.tm_sec = (setalarm.time.tm_sec + seconds) % 60;
+			setalarm.time.tm_min += 1; 
+		}
+		else {
+			setalarm.time.tm_sec += (int)seconds;
+		}
+
+		setalarm.signo = CONFIG_EXAMPLES_RTC_ALARM_SIGNO;
+		setalarm.sigvalue.sival_int = alarmid;
+
+		ret = ioctl(fd, RTC_SET_ALARM, (unsigned long)((uintptr_t)&setalarm));
```

Comment:
Before setting alarm, read the current time and store it. When callback is called, again read the current time. Check the actual time difference and print the value there.

Reply:
I have calculated the time difference and printed there in seconds.